tensorflow/core/kernels/mkl/mkl_batch_matmul_op.cc

```diff
@@ -139,22 +139,45 @@ class BatchMatMulMkl : public OpKernel {
     auto params = bmm.CreateMatMulParams(prefix, lhs.shape(), rhs.shape(),
                                          out_shape, adj_x_, adj_y_);
 
-#ifdef DNNL_AARCH64_USE_ACL
-    // ACL does not support reuse of primitives with different data.
-    // For matmul, the previous approach (PR #47775) of using Tensor addresses
-    // does not work, as the addresses are re-used in matmul with different data
-    // The counter  ensure we still benefit from caching via SetMklMatmul().
-    params->aarch64_counter =
-        MklMatMulPrimitiveFactory<float, Tlhs, Trhs,
-                                  Toutput>::IncrementCounter();
-#endif
     this->ExtendMklMatMulParams(ctx, *params);
 
     // Create or retrieve matmul primitive from cache.
     MklMatMulPrimitive<Tlhs, Trhs, Toutput>* matmul_prim =
         MklMatMulPrimitiveFactory<float, Tlhs, Trhs, Toutput>::Get(
             *params, false /* value for do_not_cache */);
 
+    Trhs* weight_data = const_cast<Trhs*>(rhs.flat<Trhs>().data());
```

Comment:
@TensorFlow-MKL This change affects the x86 build. Please let us know if this is okay. Otherwise, we can wrap it with `#ifdef DNNL_AARCH64_USE_ACL`

Reply:
I meant to highlight lines 149-210, not just a single line. Sorry.

Reply:
Thanks @penpornk for the review.

@TensorFlow-MKL in our tests we have noticed that oneDNN primitive matmul can request that weights parameter be in some other format apart from oihw (or transposed) such as in a blocked format so we check whether that is the case and do reorder if the format needs changing. If the format that primitive expects is the same reorder will not be created and executed so that hopefully should not have any effect on your use cases if reorder is not required.

Reply:
@milpuz01 I am interested to know the cases in your testing that triggered a reorder. In this primitive we create the memory::desc using strides which should be equivalent to what you are doing here.

Reply:
@penpornk I am not still able to understand why this is needed here. The earlier implementation is more generic. Can we have this under ifdef until we have a valid case where the reorder occurs?

Reply:
> @milpuz01 I am interested to know the cases in your testing that triggered a reorder. In this primitive we create the memory::desc using strides which should be equivalent to what you are doing here.

Hi @Srini511,

Sorry for the late reply.

When calling oneDNN matmul primitive some of the kernels that are used have blocking format such as `abDc8d`. This is line from oneDNN verbose log that shows that implementation of oneDNN matmul primitive assumes that weights must be in that format:

```
onednn_verbose,exec,cpu,matmul,gemm:acl,undef,src_f32::blocked:abcd:f0 wei_f32::blocked:abDc8d:f0 dst_f32::blocked:abcd:f0,attr-scratchpad:user ,,1x16x384x384:1x16x384x64:1x16x384x64
````

Before this primitive is executed we need to reorder weights (`wei_f32`) to that blocking format i.e. the following reorder must be before the execution of the above primitive:

```
onednn_verbose,exec,cpu,reorder,jit:uni,undef,src_f32::blocked:abcd:f0 dst_f32::blocked:abDc8d:f0,,,1x16x384x64
```

Without the change in this pull request this order would not be inserted. if the oneDNN matmul primitive wants weights in `abcd` format then the code in this pull request for reorder will have no effect.

Hope this clarifies why we need reorder and if there is a better place where to do this please do let us know.

Best,
Milos.

Reply:
Hi @Srini511,

> In the implementation here, we have created the `memory::desc` and `primitive` using strides which matches exactly with your format specification. That is why I thought the check should always fail. Let me see if I can reproduce this on my end.

Yes, that is right. Thanks for checking.

> These reorders can also incur sizeable overheads and thats why we usually cache the reordered weights (for realtime inference).

That is very true. I guess probably similar code to one here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/mkl/mkl_matmul_op_fused.cc#L230-L257 should be also added in this code so that reorders are cached first time they get called if inference is being run (`AreWeightsFrozen()` is true)

Reply:
@penpornk @milpuz01 can we add an ifdef around the block or remove the block?

Reply:
@Srini511 @milpuz01 I've applied the guard in 034f6df3318af182a7fa6f853a9bdf3b9e8d6da6. Please continue the discussion here until we agree on whether the block should be removed. Thank you!